File: Java/KvK.java
package com.company;

import java.io.*;
import java.util.*;

public class KvK{
    private Scanner fileReader;
    private String filepath;
    private File file;
    private int i;
    private String[] currentLine;
    private String lineString;

    KvK(String filePath){
        filepath = filePath;
        if(filepath.subString((filepath.length())-4, filepath.length()) != ".kvk"){
            throw new Exception("File extention must be \".kvk\"");
        }
        file = new File(filePath);
    }

    public ArrayList<Map<String, Map<String, String>>> read() throws Exception{
        fileReader = new Scanner(file);
        Boolean BOF = false;
        Boolean EOF = false;
        ArrayList<Map<String, Map<String, String>>> res = new ArrayList<Map<String, Map<String, String>>>(); // total
        String start = String.valueOf(((fileReader.nextLine()).split(" "))[0]);
        String SOF = "<#";
        if (start.equals(SOF)) {
            BOF = true;
            String line = fileReader.nextLine();
            lineString = line;
            currentLine = line.split(" ");
            while(/*fileReader.hasNextLine()*/EOF == false){
                if(lineString == "\n"){
                    line = fileReader.nextLine();
                    lineString = line;
                    currentLine = line.split(" ");
                }
                i = 0;
                while(i<currentLine.length){
                    while (currentLine[i].equals("\t\t") || currentLine[i].equals(" ") || currentLine[i].equals("\n") || currentLine[i].equals("\t") || currentLine[i].equals("")) {
                        i += 1;
                        System.out.println("skip in read");
                    }
                    if(currentLine[i].equals("class")){
                        System.out.println("class");
                        res.add(__getClass__());
                        System.out.println("class added");
                        break;
                    }
                    /**/
                    if(currentLine[0].equals("#>")){
                        EOF = true;
                        System.out.println("EOF found");
                        break;
                    }
                }
            }
        }
        else {
            throw new Exception("Expected \"<#\" at the beginning of the file");
        }
        if(BOF == true || EOF == true) {
            return res;
        } else {
            throw new Exception("Expected \"<#\" at the beginning of the file and \"#>\" at the end of the file");
        }
    }

    public void write(ArrayList<Map<String, Map<String, String>>> content) throws Exception {
        int j, k;
        String toWrite = "<#";
        for(j=0; j<content.toArray().length; j++){ // scorre le classi
            Map<String, Map<String, String>> classDict = content.get(j);
            if(classDict.keySet().toArray().length > 1){
                throw new Exception("Only one class name is allowed for each class");
            }
            String className = classDict.keySet().toArray()[0].toString();
            toWrite += "\n\tclass \"" + className + "\" ::>";
            Map<String, String> attrsMap = classDict.get((classDict.keySet()).toArray()[0]);
            String[] attrs = attrsMap.keySet().toArray(new String[0]);
            for(k=0; k< attrs.length; k++){
                String attrContent = attrsMap.get(attrs[k]);
                toWrite += "\n\t\t(" + attrs[k] + ") -> \"" + attrContent + "\"";
            }
        }
        toWrite += "\n#>";
        System.out.println(toWrite);
        BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
        writer.write(toWrite);
        writer.close();
    }

    public Map<String, String> getclass(String className){
        boolean found = false;
        KvK kvk = new KvK(filepath);
        Map<String, Map<String, String>> current = new HashMap<String, Map<String, String>>();
        try {
            ArrayList<Map<String, Map<String, String>>> content = kvk.read();
            int j = 0;
            for (j = 0; j < content.size(); j++) {
                current = content.get(j);
                if (current.containsKey(className)) {
                    found = true;
                    break;
                }
            }
            if (found) {
                Map<String, String> res = current.get(className);
                return res;
            } else {
                return null;
            }
        } catch (Exception ex){
            ex.printStackTrace();
        }
        return null;
    }

    public String getattr(String attribute, String className){
        KvK kvk = new KvK(filepath);
        boolean found = false;
        Map<String, Map<String, String>> current = new HashMap<String, Map<String, String>>();
        try{
            ArrayList<Map<String, Map<String, String>>> content = kvk.read();
            int j = 0;
            for (j = 0; j < content.size(); j++) {
                current = content.get(j);
                if (current.containsKey(className)) {
                    found = true;
                    break;
                }
            }
            if (found) {
                Map<String, String> tmp = current.get(className);
                String res = tmp.get(attribute);
                return res;
            } else {
                return null;
            }
        } catch (Exception ex){
            ex.printStackTrace();
        }
        return null;
    }

    public void addClass(String className){
        if(file.exists()){
            try {
                fileReader = new Scanner(file);
                String fileContent = fileReader.nextLine();
                while(fileReader.hasNextLine()){
                    fileContent += "\n" + fileReader.nextLine();
                }
                fileContent = fileContent.substring(0, fileContent.length()-2);
                fileContent += "    class \"" + className + "\" ::>\n#>";
                try {
                    BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
                    writer.write(fileContent);
                    writer.close();
                } catch (IOException ex){
                    ex.printStackTrace();
                }
            } catch (FileNotFoundException ex){
                ex.printStackTrace();
            }
        } else {
            String toWrite = "<#\n    class \"" + className + "\" ::>\n#>";
            try {
                BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
                writer.write(toWrite);
                writer.close();
            } catch (IOException ex){
                ex.printStackTrace();
            }
        }
    }

    public void addAttr(String className, String attrName, String attrContent) throws Exception{
        if(file.exists()){
            try {
                fileReader = new Scanner(file);
                String fileContent = fileReader.nextLine();
                while(fileReader.hasNextLine()){
                    fileContent += "\n" + fileReader.nextLine();
                }
                int classLength = ("class \"" + className + "\" ::>").length();
                int classIndex = fileContent.indexOf("class \"" + className + "\" ::>");
                if(classIndex != -1){
                    int index = classIndex+classLength;
                    int attrIndex = fileContent.indexOf("\n        (" + attrName + ") -> \"" + attrContent + "\"");
                    if(attrIndex != -1){
                        throw new Exception("Attribute already in class.");
                    }
                    String preAttr = fileContent.substring(0, index);
                    String afterAttr = fileContent.substring(index, fileContent.length());
                    String toWrite = preAttr + "\n        (" + attrName + ") -> \"" + attrContent + "\"" + afterAttr;
                    try {
                        BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
                        writer.write(toWrite);
                        writer.close();
                    } catch (IOException ex){
                        ex.printStackTrace();
                    }
                } else {
                    throw new Exception("Class not found in file");
                }

            } catch (FileNotFoundException ex){
                ex.printStackTrace();
            }
        } else {
            throw new Exception("You must create at least a class before adding atributes.");
        }
    }

    public void editClass(String className, String newClasName) throws Exception{
        fileReader = new Scanner(file);
        String fileContent = fileReader.nextLine();
        while(fileReader.hasNextLine()){
            fileContent += "\n" + fileReader.nextLine();
        }
        int startIndex = (fileContent.indexOf("class \"" + className + "\" ::>"));
        if(startIndex == -1){
            throw new Exception("Class not found");
        }
        startIndex += 7;
        int endIndex = startIndex+className.length();
        fileContent = fileContent.substring(0,startIndex) + newClasName + fileContent.substring(endIndex,fileContent.length());
        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
            writer.write(fileContent);
            writer.close();
        } catch (IOException ex){
            ex.printStackTrace();
        }
    }

    public void editAttr(String className, String attrName, String newAttrName) throws Exception{
        fileReader = new Scanner(file);
        String fileContent = fileReader.nextLine();
        while(fileReader.hasNextLine()){
            fileContent += "\n" + fileReader.nextLine();
        }
        int startIndex = (fileContent.indexOf("class \"" + className + "\" ::>"));
        if(startIndex == -1){
            throw new Exception("Class not found");
        }
        String preclass = fileContent.substring(0, startIndex);
        String tmp = fileContent.substring(startIndex+5, fileContent.length());
        int endClassIndex = tmp.indexOf("class");
        if (endClassIndex == -1){
            endClassIndex = tmp.indexOf("#>");
        }
        String isolatedClass = fileContent.substring(startIndex, endClassIndex+startIndex+5);
        String afterClass = tmp.substring(endClassIndex, tmp.length());

        int attrNameIndex = isolatedClass.indexOf("(" + attrName + ")");
        if (attrNameIndex == -1){
            throw new Exception(("Attribute not found"));
        }
        isolatedClass = isolatedClass.substring(0, attrNameIndex+1) + newAttrName +
                isolatedClass.substring(attrNameIndex+attrName.length()+1, isolatedClass.length());
        fileContent = preclass + isolatedClass + afterClass;
        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
            writer.write(fileContent);
            writer.close();
        } catch (IOException ex){
            ex.printStackTrace();
        }
    }

    public void editAttrContent(String className, String attrName, String newAttrContent) throws Exception{
        fileReader = new Scanner(file);
        String fileContent = fileReader.nextLine();
        while(fileReader.hasNextLine()){
            fileContent += "\n" + fileReader.nextLine();
        }
        int startIndex = (fileContent.indexOf("class \"" + className + "\" ::>"));
        if(startIndex == -1){
            throw new Exception("Class not found");
        }
        String preclass = fileContent.substring(0, startIndex);
        String tmp = fileContent.substring(startIndex+5, fileContent.length());
        int endClassIndex = tmp.indexOf("class");
        if (endClassIndex == -1){
            endClassIndex = tmp.indexOf("#>");
        }
        String isolatedClass = fileContent.substring(startIndex, endClassIndex+startIndex+5);
        String afterClass = tmp.substring(endClassIndex, tmp.length());

        int attrNameIndex = isolatedClass.indexOf("(" + attrName + ")");
        if (attrNameIndex == -1){
            throw new Exception(("Attribute not found"));
        }
        int attrIndex = attrNameIndex;
        while(!(isolatedClass.substring(attrIndex, attrIndex+1)).equals("\"")){
            attrIndex += 1;
        }
        int endAttrIndex = attrIndex+1;
        while(!(isolatedClass.substring(endAttrIndex, endAttrIndex+1)).equals("\"")){
            endAttrIndex += 1;
        }
        isolatedClass = isolatedClass.substring(0, attrIndex+1) + newAttrContent + isolatedClass.substring(endAttrIndex, isolatedClass.length());
        fileContent = preclass + isolatedClass + afterClass;
        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
            writer.write(fileContent);
            writer.close();
        } catch (IOException ex){
            ex.printStackTrace();
        }
    }

    public void removeClass(String className) throws Exception{
        fileReader = new Scanner(file);
        String fileContent = fileReader.nextLine();
        while(fileReader.hasNextLine()){
            fileContent += "\n" + fileReader.nextLine();
        }
        int classIndex = fileContent.indexOf("class \"" + className + "\" ::>");
        if(classIndex == -1){
            throw new Exception("Class not found.");
        }
        String preclass = fileContent.substring(0, classIndex-4);
        String tmp = fileContent.substring(classIndex+5, fileContent.length());
        int endClassIndex = tmp.indexOf("class");
        String afterClass;
        if(endClassIndex == -1){
            endClassIndex = tmp.indexOf("#>");
            afterClass = "#>";
        } else {
            afterClass = tmp.substring(endClassIndex, tmp.length());
        }
        fileContent = preclass + afterClass;
        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
            writer.write(fileContent);
            writer.close();
        } catch (IOException ex){
            ex.printStackTrace();
        }
    }

    public void removeAttr(String className, String attrName) throws Exception{
        fileReader = new Scanner(file);
        String fileContent = fileReader.nextLine();
        while(fileReader.hasNextLine()){
            fileContent += "\n" + fileReader.nextLine();
        }
        int startIndex = (fileContent.indexOf("class \"" + className + "\" ::>"));
        if(startIndex == -1){
            throw new Exception("Class not found");
        }
        String preclass = fileContent.substring(0, startIndex);
        String tmp = fileContent.substring(startIndex+5, fileContent.length());
        int endClassIndex = tmp.indexOf("class");
        if (endClassIndex == -1){
            endClassIndex = tmp.indexOf("#>");
        }
        String isolatedClass = fileContent.substring(startIndex, endClassIndex+startIndex+1);
        String afterClass = tmp.substring(endClassIndex, tmp.length());

        int attrNameIndex = isolatedClass.indexOf("(" + attrName + ")");
        if (attrNameIndex == -1){
            throw new Exception(("Attribute not found"));
        }
        int attrIndex = attrNameIndex;
        while(!(isolatedClass.substring(attrIndex, attrIndex+1)).equals("\"")){
            attrIndex += 1;
        }
        int endAttrIndex = attrIndex+1;
        while(!(isolatedClass.substring(endAttrIndex, endAttrIndex+1)).equals("\"")){
            endAttrIndex += 1;
        }
        isolatedClass = isolatedClass.substring(0, attrNameIndex-8) + isolatedClass.substring(endAttrIndex+1, isolatedClass.length()) + "    ";
        fileContent = preclass + isolatedClass + afterClass;
        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(filepath));
            writer.write(fileContent);
            writer.close();
        } catch (IOException ex){
            ex.printStackTrace();
        }
    }

    private Map<String, Map<String, String>> __getClass__() throws Exception{
        Map<String, Map<String, String>> tmpRes = new HashMap<String, Map<String, String>>(); // classes
        String className = new String();
        i += 1;
        className = currentLine[i];
        className = className.substring(1, (className.length())-1);
        System.out.println("className: " + className);
        i += 1;
        if(!currentLine[i].equals("::>")){
            throw new Exception("Expected \"::>\" after class declaration");
        }
        String line = fileReader.nextLine();
        lineString = line;
        currentLine = line.split(" ");
        while (currentLine[i].equals("\t\t") || currentLine[i].equals(" ") || currentLine[i].equals("\n") || currentLine[i].equals("\t") || currentLine[i].equals("") || currentLine[i].equals("\t\t")) {
            i += 1;
            System.out.println("skip in class");
        }
        if(((currentLine[i]).substring(0, 1)).equals("(")){
            while (currentLine[i].equals("\t\t") || currentLine[i].equals(" ") || currentLine[i].equals("\n") || currentLine[i].equals("\t") || currentLine[i].equals("")) {
                i += 1;
            }
            System.out.println("attr");
            tmpRes.put(className, __getAttr__());
            System.out.println("attrs received");
        }
        return tmpRes;
    }

    private Map<String, String> __getAttr__() throws Exception {
        Map<String, String> stringMap = new HashMap<String, String>();
        while(((currentLine[i]).substring(0, 1)).equals("(")) {
            if (!(currentLine[i]).substring((currentLine[i].length()) - 1, (currentLine[i].length())).equals(")")) {
                throw new Exception("Expected \")\" after attribute name declaratino");
            }
            String attrName = (currentLine[i]).substring(1, (currentLine[i]).length() - 1);
            System.out.println("attrName: " + attrName);
            i += 1;
            if (currentLine[i].equals("->")) {
                i += 1;
                String attrContent = (currentLine[i]).substring(1, (currentLine[i]).length() - 1);
                stringMap.put(attrName, attrContent);
                System.out.println("attrContent: " + attrContent);
            } else {
                throw new Exception("Expected \"(\" before attribute name declaration");
            }
            i = 0;
            String line = fileReader.nextLine();
            lineString = line;
            currentLine = line.split(" ");
            while (currentLine[i].equals("\t\t") || currentLine[i].equals(" ") || currentLine[i].equals("\n") || currentLine[i].equals("\t") || currentLine[i].equals("")) {
                i += 1;
                System.out.println("a");
            }
            System.out.println("end of attr mainloop");
        }
        return stringMap;
    }
}

File: C++/KvK.cpp
//
// Created by Cargo on 12/05/2022.
//

#include "KvK.h"

void KvK::startingContent() {
    if (subString(this->filePath, this->filePath.length() -4, this->filePath.length()) != ".kvk") // checks the extension of the file
        throw runtime_error("File extension must be \".kvk\"");
    else {
        file.open(this->filePath, ios::in); // tries to open file to read the content
        if(file.is_open()){
            string fileContent;
            while(getline(file, fileContent)){
                this->content += fileContent + "\n"; // reads the content and stores it in "content"
            }
            file.close();
        } else { // if it is impossible to read
            this->content = "<#\n#>"; // sets content to the empty standard
            file.open(this->filePath, ios::out);
            if(file.is_open()){
                file << this->content; // writes the standard empty content in the file
                file.close();
            }
        }
    }
}

map<string, map<string, string>> KvK::getClass_() {
    map<string, map<string, string>> tmpDict; // content of the class variable

    this->pos += 5;

    while(this->text[this->pos] == ' ') // skips blank spaces
        this->pos += 1;

    if(this->text[this->pos] == '"'){ // finds the opening double quotes for class name
        this->pos += 1;
        string className;

        while(this->text[this->pos] != '"'){ // scans until the following double quotes to read class name
            className += this->text[this->pos];
            this->pos += 1;
        }
        this->pos += 1;

        while(this->text[this->pos] == ' ') // skips blank spaces
            this->pos += 1;

        map<string, string> insideDict; // map for attributes
        if(this->text.substr(this->pos, 3) == "::>"){ // looks for the attribute list opening tag
            this->pos += 3;

            while(subString(this->text, this->pos, this->text.length()-1) != "#>"
            and this->text.substr(this->pos, 5) != "class"){ // scans until it finds EOF or new class
                this->nullAdvancement(); // skips blank spaces, blank lines and tabs

                if(this->text[this->pos] == '('){ // looks for attribute opening parentheses
                    vector<string> returnVector = this->getAttr_(); // starts attribute reading, one at the time
                    insideDict.insert(pair<string, string>(returnVector[0], returnVector[1])); // stores the attribute in the map
                    this->pos += 1;
                }

                tmpDict.insert(pair<string, map<string, string>>(className, insideDict)); // pushes the class in the return map
            }
            return tmpDict;
        } else
            throw runtime_error("\"::>\" must follow class name");
    } else
        throw runtime_error("Class name bust be wrapped with \"\"");
}

vector<string> KvK::getAttr_() {
    vector<string> res; // return variable
    string attrName;

    this->pos += 1;
    this->nullAdvancement(); // skips blank spaces, blank lines and tabs

    while(this->text[this->pos] != ')'){ // looks for the ending attribute name tag
        attrName += this->text[this->pos];
        this->pos += 1;
    }

    this->pos += 1;
    this->nullAdvancement(); // skips lbank lines, blank spaces and tabs

    if(this->text.substr(this->pos, 2) == "->"){ // looks for the appartenence tag
        this->pos += 2;

        while(this->text[this->pos] == ' ') // skips blank spaces
            this->pos += 1;

        if(this->text[this->pos] == '"'){ // looks for the attribute content opening double quote
            string attr;
            this->pos += 1;

            while(this->text[this->pos] != '"'){ // looks for the attribute content closing double quote
                attr += this->text[this->pos];
                this->pos += 1;
            }

            res.push_back(attrName);
            res.push_back(attr);

            return res; // returns attribute name and content
        } else
            throw runtime_error("Attribute must be wrapped with \"\"");
    } else
        throw runtime_error("\"->\" must between attribute name and attribute");
}

vector<map<string, map<string, string>>> KvK::read() {
    this->pos = 0;
    file.open(this->filePath, ios::in);
    if(file.is_open()) { // reads the content of the file
        string fileContent;
        while (getline(file, fileContent)) {
            this->text += fileContent + "\n";
        }

        file.close();

        vector<map<string, map<string, string>>> res; // return variable
        if (this->text.substr(this->pos, 2) == "<#"){ // looks for SOF
            this->pos += 2;
            while(this->pos < this->text.length()
            and subString(this->text, this->pos, this->text.length()-1) != "#>"){ // scans until the end of the file
                this->nullAdvancement(); // skips blank spaces, blank lines and tabs

                if(this->text.substr(this->pos, 5) == "class"){ // looks for the class tag
                    map<string, map<string, string>> point = getClass_(); // starts the class reding process
                    res.push_back(point); // stores the class
                }
                if(subString(this->text, this->pos, this->text.length()-1) == "#>") // if it find the EOF, stops scan
                    break;
            }
            return res;
        } else {
            throw runtime_error("File not starting with <#");
        }
    } else
        throw runtime_error("Can't open file \"" + this->filePath + "\"");
}

void KvK::write(const vector<map<string, map<string, string>>> &toWriteContent) {
    string toWrite = "<#\n"; // SOF

    for(auto& classcont:toWriteContent){ // for each class container in the vector
        for(auto& internalClass: classcont){ // for each pair in class container
            auto className = internalClass.first; // className is the first element of the pair
            toWrite += "    class \"" + className + "\" ::>\n";
            auto attributeList = internalClass.second; // map of attributes
            for(auto& attribute : attributeList) { // gets pair from map
                auto attributeName = attribute.first;
                auto attributeContent = attribute.second; // gets the attribute content, which is the second element of the pair
                toWrite += "        (" + attributeName + ") -> \"" + attributeContent + "\"\n";
            }
        }
        toWrite += "#>"; // EOF
    }
    file.open(this->filePath, ios::out);
    if(file.is_open()){ // writes the content in the file
        file << this->content;
        file.close();
    }
}

map<string, string> KvK::getClass(const string &className) {
    this->pos = 0;
    auto fileContent = this->read(); // reads the file

    try {
        for(const auto& classCont:fileContent){ // scans the class containers
            for(const auto& classname:classCont){ // scans teh classes
                if(classname.first == className){ // if the class name is equal to the wanted one
                    return classname.second; // return sthe class content
                }
            }
        }
    } catch (...){
        map<string, string> empty;
        return empty;
    }
    map<string, string> empty;
    return empty;
}

string KvK::getAttr(const string &className, const string &attrName) {
    this->pos = 0;
    auto fileContent = this->read();

    try {
        for(const auto& classCont:fileContent){ // scans the class containers
            for(const auto& classname:classCont){ // scans the classes
                if(classname.first == className){ // if the class name is equal to the wanted one
                    auto attrs = classname.second;
                    for(const auto& attr:attrs){ // scans the attributes
                        if(attr.first == attrName) // if the attribute name is equal to the wanted one
                            return attr.second; // returns the attribute content
                    }
                }
            }
        }
    } catch (...){
        return "";
    }
    return "";
}

bool KvK::addClass(const string &className) {
    int classIndex = this->content.find("    class \"" + className + "\" ::>");
    if(classIndex < 0){ // cheks if the class already exists
        string oldContent = subString(this->content, 0, this->content.length()-3); // takes teh content but removes the EOF
        string eof = "#>";

        this->content = oldContent + "    class \"" + className + "\" ::>\n" + eof; // adds the new class and then the EOF

        file.open(this->filePath, ios::out);
        if(file.is_open()){ // writes on the file the new content
            file << this->content;
            file.close();
            return true;
        }
        return false;
    }
    return false;
}

bool KvK::addAttr(const string &className, const string &attrName, const string &attrContent) {
    int index = this->content.find("class \"" + className + "\" ::>");
    if (index > 0) { // checks if the class exists
        int endIndex = index + ("class \"" + className + "\" ::>").length()+1; // takes teh index of the class

        int endClassIndex = endIndex+1;
        try { // scans until EOF or new class
            while (this->content.substr(endClassIndex, 2) != "#>"
                   and this->content.substr(endClassIndex,5) != "class")
                endClassIndex += 1;
        } catch(...) {
            endClassIndex = this->content.length()-2;
        }

        string isolatedClass = subString(this->content, index, endClassIndex); // isolates the wanted class
        int attrIndex = isolatedClass.find("        (" + attrName + ") -> \"" + attrContent + "\"");
        if(attrIndex < 0) { // checks if attribute already exists
            this->content = subString(this->content, 0, endIndex) + "        (" + attrName + ") -> \""
                    + attrContent + "\"\n"
                    + subString(this->content, endIndex, this->content.length()); // adds the attribute

            file.open(this->filePath, ios::out);
            if (file.is_open()) { // writes on the file the new content
                file << this->content;
                file.close();

                return true;
            }
            return false;
        }
    } else
        throw ClassNotFoundError("Class not found");
}

bool KvK::editClass(const string &className, const string &newClassName) {
    int startIndex = this->content.find("class \"" + className + "\" ::>");
    if(startIndex > 0) { // checks if the class exists
        int endIndex = startIndex + ("class \"" + className + "\" ::>").length(); // takes the ending index of teh class declaration

        this->content = this->content.substr(0, startIndex) + "class \"" + newClassName + "\" ::>"
                + subString(this->content, endIndex, this->content.length()); // edits the class name

        file.open(this->filePath, ios::out);
        if(file.is_open()){ // writes the new content on the file
            file << this->content;
            file.close();

            return true;
        }
        return false;
    } else
        throw ClassNotFoundError("Class \"" + className + "\" not found");
}

bool KvK::editAttr(const string &className, const string &attrName, const string &newAttrName, const string &attrContent) {
    int classIndex = this->content.find("class \"" + className + "\" ::>");
    if (classIndex > 0) { // checks if the class exists
        string preClass = this->content.substr(0, classIndex); // isolates what's before the class
        string tmp = subString(this->content, classIndex+5, this->content.length());

        int eofClassIndex = tmp.find("class");
        int endClassIndex = tmp.find("#>");

        int endIndex;
        if(eofClassIndex > 0)
            endIndex = eofClassIndex;
        else if(endClassIndex > 0)
            endIndex = endClassIndex;

        string isolatedClass = subString(this->content, classIndex, endClassIndex+classIndex+5); // isolates the class
        string afterClass = tmp.substr(endClassIndex, tmp.length()); // isolates what's after the class

        int oldAttrIndex = isolatedClass.find(attrName);
        if(oldAttrIndex > 0) { // checks if the attri bute exists
            isolatedClass = isolatedClass.substr(0, oldAttrIndex) + newAttrName + // updates the attribute name
                            subString(isolatedClass, oldAttrIndex+attrName.length(), isolatedClass.length());

            if(not attrContent.empty()){ // if attribute name is not empty
                int virgStart = oldAttrIndex;

                while(isolatedClass[virgStart] != '"')
                    virgStart += 1;

                int virgEnd = virgStart+1;
                while(isolatedClass[virgEnd] != '"')
                    virgEnd += 1;

                isolatedClass = isolatedClass.substr(0, virgStart+1) + attrContent // updates the attribute content
                                    + subString(isolatedClass, virgEnd, isolatedClass.length());
            }
            this->content = preClass + isolatedClass + afterClass;

            file.open(this->filePath, ios::out);
            if(file.is_open()){ // writes the new content in the file
                file << this->content;
                file.close();

                return true;
            }
            return false;

        } else
            throw AttributeNotFoundError("Attribute \"" + attrName + "\" of class \""
                                                                + className + "\" not found");
    } else
        throw ClassNotFoundError("Class \"" + className + "\" not found");
}

bool KvK::removeClass(const string &className) {
    int classIndex = this->content.find("    class \"" + className + "\" ::>");
    if (classIndex > 0) { // checks if the class exists
        string preClass = subString(this->content, 0, classIndex); // isolates what's before the class
        string tmp = subString(this->content, classIndex+5, this->content.length());

        int endClassIndex = tmp.find("class");
        int eofClassIndex = tmp.find("#>");

        int endIndex;
        string afterClass;

        if(endClassIndex > 0 and endClassIndex < eofClassIndex) // isolates what's after the class
            afterClass = subString(tmp, endClassIndex-4, tmp.length());
        else if(eofClassIndex > 0)
            afterClass = "#>";

        this->content = preClass + afterClass; // updates the content

        file.open(this->filePath, ios::out);
        if(file.is_open()){ // writes the new content in the file
            file << this->content;
            file.close();

            return true;
        }
        return false;
    } else
        throw ClassNotFoundError("Class \"" + className + "\" not found");
}

bool KvK::removeAttr(const string &className, const string &attrName) {
    int classIndex = this->content.find("class \"" + className + "\" ::>");
    if (classIndex > 0) { // checks if the class exists
        string preClass = subString(this->content, 0, classIndex); // isolates what's after the class
        string tmp = subString(this->content, classIndex+5, this->content.length());

        int eofClassIndex = tmp.find("#>");
        int endClassIndex = tmp.find("class");

        int endIndex;

        if(endClassIndex > 0 and endClassIndex < eofClassIndex)
            endIndex = endClassIndex;
        else
            endIndex = eofClassIndex;

        string isolatedClass = subString(this->content, classIndex, endClassIndex+classIndex+5); // isolates the class
        string afterClass = subString(tmp, endClassIndex, tmp.length()); // isolates what's after the class

        int oldAttrIndex = isolatedClass.find("(" + attrName + ")");

        if(oldAttrIndex >= 0) { // checks if the attribute exists
            int virgStart = oldAttrIndex;
            while(isolatedClass[virgStart] != '"')
                virgStart += 1;

            int virgEnd = virgStart+1;
            while(isolatedClass[virgEnd] != '"')
                virgEnd += 1;

            isolatedClass = subString(isolatedClass, 0, oldAttrIndex) // updates the class
                          + subString(isolatedClass, virgEnd+1, isolatedClass.length());

            this->content = preClass + isolatedClass + afterClass;

            file.open(this->filePath, ios::out);
            if(file.is_open()){ // writes the new content in the file
                file.clear();
                file << this->content;
                file.close();

                return true;
            }
            return false;

        } else
            throw AttributeNotFoundError("Attribute \"" + attrName + "\" of class \""
                                         + className + "\" not found");
    } else
        throw ClassNotFoundError("Class \"" + className + "\" not found");
}

bool KvK::isEmpty() {
    string actualContent = this->content;
    auto contentList = split(actualContent, ' '); // splits the content, removing blank spaces

    int i = 0;
    for(const auto& element:contentList){ // removes all the blank speces
        if(element.empty() or element == " "){
            contentList = pop(contentList, i);
        }
        i += 1;
    }
    try{ // checks if the file is litterally empty, or contains the empty standard
        if(contentList[0] == "<#" and contentList[1] == "#>")
            return true;
        return false;
    } catch (...) {
        if(contentList.empty())
            return true;
        return false;
    }
}